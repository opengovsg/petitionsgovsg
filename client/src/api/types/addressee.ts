import { BaseModelParams, MessageResponse } from './common'

export type BaseAddresseeDto = BaseModelParams & {
  name: string
  shortName: string | null
}

export type CreateAddresseeReqDto = {
  name: string
  shortName: string | null
}

// Returns signature ID
export type CreateAddresseeResDto = MessageResponse & { data: number }
