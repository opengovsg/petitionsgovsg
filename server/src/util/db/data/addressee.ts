export const mockAddressee = {
  name: 'Ministry of Environment',
  shortName: 'MoE',
}

export const mockAddressees = [
  {
    id: 1,
    name: 'Ministry of Environment',
    shortName: 'MoE',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
    openToPublic: false,
    officerEmails: [],
  },
  {
    id: 2,
    name: 'Ministry of Labour',
    shortName: 'MoL',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
    openToPublic: false,
    officerEmails: [],
  },
]
