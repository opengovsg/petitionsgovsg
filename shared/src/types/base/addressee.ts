import { z } from 'zod'
import { BaseModel } from './common'

export const Addressee = BaseModel.extend({
  name: z.string(),
  shortName: z.string().nullable(),
  openToPublic: z.boolean(),
  officerEmails: z.array(z.string()).nullable(),
})

export type Addressee = z.infer<typeof Addressee>
