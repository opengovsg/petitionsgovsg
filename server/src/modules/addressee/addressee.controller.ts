import { StatusCodes } from 'http-status-codes'
import { createLogger } from '../../bootstrap/logging'
import { Message } from '../../types/message-type'
import { ControllerHandler } from '../../types/response-handler'
import { AddresseeDto, AddresseeService } from './addressee.service'

const logger = createLogger(module)

export class AddresseeController {
  private addresseeService: Public<AddresseeService>

  constructor({
    addresseeService,
  }: {
    addresseeService: Public<AddresseeService>
  }) {
    this.addresseeService = addresseeService
  }

  /**
   * Lists all post
   * @query sort Sort by popularity or recent
   * @query size Number of posts to return
   * @query page If size is given, specify which page to return
   * @return 200 with posts and totalItem for pagination
   * @return 500 when database error occurs
   */
  listAddressees: ControllerHandler<
    undefined,
    AddresseeDto[] | Message,
    undefined,
    undefined
  > = async (req, res) => {
    try {
      const data = await this.addresseeService.listAddressees()
      return res.status(StatusCodes.OK).json(data)
    } catch (error) {
      if (error instanceof Error) {
        logger.error({
          message: 'Error while listing addresees',
          meta: {
            function: 'listAddressees',
          },
          error,
        })
        return res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Server Error' })
      }
    }
  }
}
