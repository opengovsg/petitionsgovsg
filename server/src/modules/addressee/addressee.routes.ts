import express from 'express'
import { AddresseeController } from './addressee.controller'

export const routeAddressees = ({
  controller,
}: {
  controller: AddresseeController
}): express.Router => {
  const router = express.Router()
  /**
   * Lists all addressees
   * @route   GET /api/posts
   * @returns 200 with posts
   * @returns 500 when database error occurs
   * @access  Public
   */
  router.get('/', controller.listAddressees)
  return router
}
