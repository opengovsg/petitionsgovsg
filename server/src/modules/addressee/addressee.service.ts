import { Addressee } from '~shared/types/base'
import { ModelDef } from '../../types/sequelize'

export type AddresseeDto = Pick<Addressee, 'id' | 'name' | 'shortName'>

export class AddresseeService {
  private Addressee: ModelDef<Addressee>

  constructor({ Addressee }: { Addressee: ModelDef<Addressee> }) {
    this.Addressee = Addressee
  }

  /**
   * Lists all addressees
   */
  listAddressees = async (): Promise<AddresseeDto[]> => {
    const addressees = (await this.Addressee.findAll({
      attributes: ['id', 'name', 'shortName'],
    })) as AddresseeDto[]
    return addressees
  }
}
