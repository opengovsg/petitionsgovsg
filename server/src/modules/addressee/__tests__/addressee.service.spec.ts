import {
  createTestDatabase,
  resetAndSetupDb,
  SequelizeWithModels,
} from '@/util/db/jest-db'
import { AddresseeService } from '../addressee.service'

describe('AddresseeService', () => {
  let service: AddresseeService
  let db: SequelizeWithModels

  beforeAll(async () => {
    db = await createTestDatabase()
    service = new AddresseeService({ Addressee: db.Addressee })
  })

  beforeEach(async () => {
    await resetAndSetupDb(db)
    jest.resetAllMocks()
  })

  describe('listAddressees', () => {
    it('returns all addressees', async () => {
      const addressees = await service.listAddressees()

      expect(addressees).toMatchSnapshot()
    })
  })
})
