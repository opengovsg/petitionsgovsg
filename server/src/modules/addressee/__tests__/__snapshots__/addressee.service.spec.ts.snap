// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddresseeService listAddressees returns all addressees 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Ministry of Environment",
    "shortName": "MoE",
  },
  Object {
    "id": 2,
    "name": "Ministry of Labour",
    "shortName": "MoL",
  },
]
`;
